feat(config): inheritConfig (#27864)
renovatebot/renovate: docs/usage/self-hosted-configuration.md

Change this setting to `true` to include repositories that are mirrors as Renovate targets.

## inheritConfig

When this option is enabled, Renovate will look for a file `org-inherited-config.json5` in the `<orgName>/renovate-config` repo before processing a repository, and read this in as config.
If the repository is in a nested org/group on a supported platform such as GitLab, such as `topGroup/nestedGroup/projectName` then Renovate will look in `topGroup/nestedGroup/renovate-config`.

If `inheritConfig` is `true` but the inherited config file does not exist then Renovate will proceed without warning.
If the file exists but cannot be parsed, then a config warning issue will be raised and the job aborted.

Inherited config can include all valid repo config and additionally the following config options:

- `requireConfig`
- `onboarding`
- Any other global option starting with `onboarding`, such as `onboardingBranch`

> // TODO: confirm if there are any others config options useful/applicable

Author: TODO

If `inheritConfig` is `true` but the inherited config file does not exist then Renovate will proceed without warning.
If the file exists but cannot be parsed, then a config warning issue will be raised and the job aborted.

Inherited config can include all valid repo config and additionally the following config options:

- `requireConfig`
- `onboarding`
- Any other global option starting with `onboarding`, such as `onboardingBranch`

// TODO: confirm if there are any others config options useful/applicable

This means that orgs can change/control default behavior for whether configs are required and how repositories are onboarded.

This setting is not enabled in the Mend Renovate App because each 404 from the GitHub API would count as a used API call, which would add millions of wasted API calls per week.
A smart/dynamic approach will be added in future so that it can be enabled selectively per-org.

## inheritConfigFileName

Change this setting if you wish for Renovate to look for a different file name within the `inheritConfigRepoName` repository.

> // TODO: check if nested works

Author: TODO

Change this setting to `true` to include repositories that are mirrors as Renovate targets.

## inheritConfig

> When this option is enabled, Renovate will look for a file `org-inherited-config.json5` in the `<orgName>/renovate-config` repo before processing a repository, and read this in as config.

Reviewer: ```suggestion
When you enable this option, Renovate will look for a file `org-inherited-config.json5` in the `<orgName>/renovate-config` repo before processing a repository, and read this in as config.
```

> If the repository is in a nested org/group on a supported platform such as GitLab, such as `topGroup/nestedGroup/projectName` then Renovate will look in `topGroup/nestedGroup/renovate-config`.

Reviewer: ```suggestion
If the repository is in a nested organization/group on a supported platform such as GitLab, such as `topGroup/nestedGroup/projectName` then Renovate will look in `topGroup/nestedGroup/renovate-config`.
```

> If `inheritConfig` is `true` but the inherited config file does not exist then Renovate will proceed without warning.

Reviewer: ```suggestion
If `inheritConfig` is `true` but the inherited config file does _not_ exist then Renovate will proceed without warning.
```

If the file exists but cannot be parsed, then a config warning issue will be raised and the job aborted.

> Inherited config can include all valid repo config and additionally the following config options:

Reviewer: ```suggestion
The inherited config may include all valid repo config and these config options:
```

- `bbUseDevelopmentBranch`
- `onboarding`
- `onboardingBranch`
- `onboardingCommitMessage`
- `onboardingConfig`
- `onboardingConfigFileName`
- `onboardingNoDeps`
- `onboardingPrTitle`
- `onboardingRebaseCheckbox`
- `requireConfig`

> This means that orgs can change/control default behavior for whether configs are required and how repositories are onboarded.

Reviewer: ```suggestion
This means that organizations can change/control default behavior like whether configs are required and how repositories are onboarded.
```

> This setting is not enabled in the Mend Renovate App because each 404 from the GitHub API would count as a used API call, which would add millions of wasted API calls per week.

Reviewer: ```suggestion
We disabled `inheritConfig` in the Mend Renovate App to avoid millions of wasted API calls per week.
This is because each `404` from the GitHub API counts as a used API call.
```

I think it's nice to explain the key point in the first sentence, and give more details in the second sentence.

> A smart/dynamic approach will be added in future so that it can be enabled selectively per-org.

Reviewer: ```suggestion
```

I don't know if it's smart to "promise" new features/behavior, unless you're _really_ sure you're going to add it soon?

I'll let you decide to keep/remove this line. 😉

## inheritConfigFileName

> Change this setting if you wish for Renovate to look for a different file name within the `inheritConfigRepoName` repository.

Reviewer: ```suggestion
Change this setting if you want Renovate to look for a different file name within the `inheritConfigRepoName` repository.
```

> Nested files are supported, e.g. `"some-dir/config.json"`;

Reviewer: ```suggestion
You may use nested files, for example: `"some-dir/config.json"`.
```

Is that `;` at the end of the current line intended? If yes, it should probably go within the `code` block.

I've removed the `;` in my suggestion, so please add it back, in the correct place if we need it. 😉

## inheritConfigRepoName

> Change this setting if you wish for Renovate to look in an alternative repository for the inherited config.

Reviewer: ```suggestion
Change this setting if you want Renovate to look in an alternative repository for the inherited config.
```

> The repository always needs to be on the same platform/endpoint, and Renovate's token needs permission to access it.

Reviewer: ```suggestion
The repository must be on the same platform and endpoint, and Renovate's token must have permission to access the repository.
```

We should use `must` to indicate requirements. 😉

Also, we should spell out what `it` refers to in the sentence. I think you mean: `it = repository`?

## inheritConfigStrict

By default Renovate will silently (debug log message only) ignore cases where `inheritConfig=true` but no inherited config is found.

> If `inheritConfigStrict=true` then Renovate will instead abort the run and raise a config error if the inherited config is not found.

Reviewer: ```suggestion
When you set `inheritConfigStrict=true` then Renovate will abort the run and raise a config error if the inherited config is not found.
```

> Set this to true only if every org has an inherited config file and you want to make sure Renovate never runs without it.

Reviewer: ```suggestion
Only set this config option to `true` if _every_ organization has an inherited config file _and_ you want to make sure Renovate _always_ uses that inherited config.
```

Let's start the sentence with the `Only` fragment first, and emphasize the key parts. I guess the key parts are:

- Only use this when EVERY organization has an inherited config file
- AND you want to make sure Renovate ALWAYS uses it

Change this setting to `true` to include repositories that are mirrors as Renovate targets.

## inheritConfig

> When you enable this option, Renovate will look for the `inheritConfigFileName` file (defaults to `org-inherited-config.json`) in the `inheritConfigRepoName` repository (defaults to `{{parentOrg}}/renovate-config`) before processing a repository, and read this in as config.
> If the repository is in a nested organization/group on a supported platform such as GitLab, such as `topGroup/nestedGroup/projectName` then Renovate will look in `topGroup/nestedGroup/renovate-config`.

Reviewer: ```suggestion
When you enable this option, Renovate will look for the `inheritConfigFileName` file in the `inheritConfigRepoName` repository before processing a repository, and read this in as config.

| Config option           | Default path                | Repository config       | Repository path                 |
| ----------------------- | --------------------------- | ----------------------- | ------------------------------- |
| `inheritConfigFileName` | `org-inherited-config.json` | `inheritConfigRepoName` | `{{parentOrg}}/renovate-config` |

If the repository is in a nested organization or group on a supported platform such as GitLab, such as `topGroup/nestedGroup/projectName` then Renovate will look in `topGroup/nestedGroup/renovate-config`.
```

We should move the "defaults to `x`" info out of the introductory sentence, because that makes the sentence very hard to read. Maybe a table like this will work for that? Or you could split the default setting into their own sentences or list.

If `inheritConfig` is `true` but the inherited config file does _not_ exist then Renovate will proceed without warning.

> If the file exists but cannot be parsed, then a config warning issue will be raised and the job aborted.

Reviewer: ```suggestion
If the file exists but cannot be parsed, then Renovate will raise a config warning issue and abort the job.
```

> The inherited config may include all valid repo config and these config options:

Reviewer: ```suggestion
The inherited config may include all valid repository config and these config options:
```

- `bbUseDevelopmentBranch`
- `onboarding`
- `onboardingBranch`
- `onboardingCommitMessage`
- `onboardingConfig`
- `onboardingConfigFileName`
- `onboardingNoDeps`
- `onboardingPrTitle`
- `onboardingRebaseCheckbox`
- `requireConfig`

> This means that organizations can change/control default behavior like whether configs are required and how repositories are onboarded.

Reviewer: ```suggestion
This way organizations can change/control the default behavior, like whether configs are required and how repositories are onboarded.
```

> We disabled `inheritConfig` in the Mend Renovate App to avoid millions of wasted API calls per week.

Reviewer: ```suggestion
We disabled `inheritConfig` in the Mend Renovate App to avoid wasting millions of API calls per week.
```

> This is because each `404` from the GitHub API for a missing org inherited config counts as a used API call.

Reviewer: ```suggestion
This is because each `404` response from the GitHub API due to a missing org inherited config counts as a used API call.
```

> A smart/dynamic approach will be added in future so that it can be enabled selectively per-org.

Reviewer: ```suggestion
We will add a smart/dynamic approach in future, so that we can selectively enable `inheritConfig` per organization.
```

If you want to commit to this work, we can have a stronger sentence like this one.

`@secustor` mentioned creating a tracking issue for this work, so people can subscribe to that issue. So if you want you can create the issue, and put it in the suggestion.

## inheritConfigFileName

Change this setting if you want Renovate to look for a different file name within the `inheritConfigRepoName` repository.
You may use nested files, for example: `"some-dir/config.json"`.

## inheritConfigRepoName

Change this setting if you want Renovate to look in an alternative repository for the inherited config.

> The repository must be on the same platform and endpoint, and Renovate's token must have permission to access the repository.

Reviewer: ```suggestion
The repository must be on the same platform and endpoint, and Renovate's token must have `write` permissions to the repository.
```

_must have permission_ is a bit vague, I think we should be clear about the level of access needed.

I guess Renovate needs `write` access?

## inheritConfigStrict

By default Renovate will silently (debug log message only) ignore cases where `inheritConfig=true` but no inherited config is found.

> When you set `inheritConfigStrict=true` then Renovate will abort the run and raise a config error if the inherited config is not found.

Reviewer: ```suggestion
When you set `inheritConfigStrict=true` then Renovate will abort the run and raise a config error if Renovate can't find the inherited config.
```

> Only set this config option to `true` if _every_ organization has an inherited config file _and_ you want to make sure Renovate _always_ uses that inherited config.

Reviewer: ```suggestion
<!-- prettier-ignore -->
!!! warning
    Only set this config option to `true` if _every_ organization has an inherited config file _and_ you want to make sure Renovate _always_ uses that inherited config.
```

Maybe we should use a admonition to emphasize this warning?

Change this setting to `true` to include repositories that are mirrors as Renovate targets.

## inheritConfig

When you enable this option, Renovate will look for the `inheritConfigFileName` file in the `inheritConfigRepoName` repository before processing a repository, and read this in as config.

If the repository is in a nested organization or group on a supported platform such as GitLab, such as `topGroup/nestedGroup/projectName` then Renovate will look in `topGroup/nestedGroup/renovate-config`.

If `inheritConfig` is `true` but the inherited config file does _not_ exist then Renovate will proceed without warning.
If the file exists but cannot be parsed, then Renovate will raise a config warning issue and abort the job.

> The inherited config may include all valid repository config and these config options:

Reviewer 2: This list may change without updating this docs, so probably hint to look for ` inheritConfigSupport` in docs table?

Author: Added note below

- `bbUseDevelopmentBranch`
- `onboarding`
- `onboardingBranch`
- `onboardingCommitMessage`
- `onboardingConfig`
- `onboardingConfigFileName`
- `onboardingNoDeps`
- `onboardingPrTitle`
- `onboardingRebaseCheckbox`
- `requireConfig`

> 

Author: ```suggestion

<!-- prettier-ignore -->
!!! note
    The above list is prepared manually and may become out of date.
    Consult the self-hosted configuration docs and look for `inheritConfigSupport` values there for the definitive list.

```
